Sources/Workspace/WorkspaceConfiguration.swift

```diff
@@ -50,39 +47,283 @@ extension Workspace {
+                if !local.isEmpty {
+                    self._mirrors.append(contentsOf: local)
+                    return
+                }
+
+                // use shared if local was not found or empty
+                let shared = try self.sharedMirrors?.mirrors() ?? [:]
+                if !shared.isEmpty {
+                    self._mirrors.append(contentsOf: shared)
+                }
+            }
+        }
+    }
+
+    public struct Mirrors {
+        private let path: AbsolutePath
+        private let fileSystem: FileSystem
+        private let deleteWhenEmpty: Bool
+
+        public init(path: AbsolutePath, fileSystem: FileSystem, deleteWhenEmpty: Bool) {
+            self.path = path
+            self.fileSystem = fileSystem
+            self.deleteWhenEmpty = deleteWhenEmpty
+        }
+
+        /// The mirrors in this configuration
+        public func mirrors() throws -> DependencyMirrors {
+            guard self.fileSystem.exists(self.path) else {
+                return DependencyMirrors()
+            }
+            return try self.fileSystem.withLock(on: self.path.parentDirectory, type: .shared) {
+                return DependencyMirrors(try Self.load(self.path, fileSystem: self.fileSystem))
+            }
+        }
+
+
+        /// Apply a mutating handler on the mirrors in this configuration
+        @discardableResult
+        public func apply(handler: (inout DependencyMirrors) throws -> Void) throws -> DependencyMirrors {
+            if !self.fileSystem.exists(self.path.parentDirectory) {
+                try self.fileSystem.createDirectory(self.path.parentDirectory, recursive: true)
+            }
+            return try self.fileSystem.withLock(on: self.path.parentDirectory, type: .exclusive) {
+                let mirrors = DependencyMirrors(try Self.load(self.path, fileSystem: self.fileSystem))
+                var updatedMirrors = DependencyMirrors(mirrors.mapping)
+                try handler(&updatedMirrors)
+                if updatedMirrors != mirrors {
+                    try Self.save(updatedMirrors.mapping, to: self.path, fileSystem: self.fileSystem, deleteWhenEmpty: self.deleteWhenEmpty)
+                }
+                return updatedMirrors
+            }
+        }
+
+        private static func load(_ path: AbsolutePath, fileSystem: FileSystem) throws -> [String: String] {
+            guard fileSystem.exists(path) else {
+                return [:]
+            }
+            let data: Data = try fileSystem.readFileContents(path)
+            let decoder = JSONDecoder.makeWithDefaults()
+            let mirrors = try decoder.decode(MirrorsStorage.self, from: data)
+            let mirrorsMap = Dictionary(mirrors.object.map({ ($0.original, $0.mirror) }), uniquingKeysWith: { first, _ in first })
+            return mirrorsMap
+        }
+
+        private static func save(_ mirrors: [String: String], to path: AbsolutePath, fileSystem: FileSystem, deleteWhenEmpty: Bool) throws {
```

Comment:
Along the same lines as https://github.com/apple/swift-package-manager/pull/3670/files#r690325561, what if we moved this to an extension on `DependencyMirrors`?

Reply:
https://github.com/apple/swift-package-manager/pull/3670/files#r690546078